src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -14640,6 +14652,228 @@ public S(A a, B b)
+        }
+
+        [Fact]
+        public void WarningOnConversion_As()
+        {
+            var source =
+@"class C
+{
+    static C F(object o)
+    {
+        return o as C;
+    }
+}";
+
+            var comp = CreateStandardCompilation(
+                new[] { source, attributesDefinitions },
+                parseOptions: TestOptions.Regular.WithNullCheckingFeature(includeNonNullChecking: false));
+            comp.VerifyDiagnostics();
+
+            comp = CreateStandardCompilation(
+                new[] { source, attributesDefinitions },
+                parseOptions: TestOptions.Regular.WithNullCheckingFeature(includeNonNullChecking: true));
+            comp.VerifyDiagnostics(
+                // (5,16): warning CS8603: Possible null reference return.
+                //         return o as C;
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "o as C").WithLocation(5, 16));
+        }
+
+        [Fact(Skip = "TODO")]
+        public void DereferenceChain()
+        {
+            var source =
+@"class Person
+{
+    internal string FirstName { get; set; }
+    internal string LastName { get; set; }
+    internal string? MiddleName { get; set; }
+}
+class Program
+{
+    static void F(Person p)
+    {
+        G(p.MiddleName);
+        if (p.MiddleName != null) G(p.MiddleName);
+    }
+    static void G(string s)
+    {
+    }
+}";
+            var comp = CreateStandardCompilation(
+                new[] { source, attributesDefinitions },
+                parseOptions: TestOptions.Regular.WithNullCheckingFeature(includeNonNullChecking: false));
+            comp.VerifyDiagnostics(
+                // (11,11): warning CS8604: Possible null reference argument for parameter 's' in 'void Program.G(string s)'.
+                //         G(p.MiddleName);
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "p.MiddleName").WithArguments("s", "void Program.G(string s)").WithLocation(11, 11));
+        }
+
+        [Fact]
+        public void LocalTypeInference()
+        {
+            var source =
+@"class C
+{
+    static void F(string? s, string? t)
+    {
+        if (s != null)
+        {
+            var x = s;
+            G(x); // no warning
+            x = t;
+        }
+        else
+        {
+            var y = s;
+            G(y); // warning
+            y = t;
+        }
+    }
+    static void G(string s)
+    {
+    }
+}";
+            var comp = CreateStandardCompilation(
+                new[] { source, attributesDefinitions },
+                parseOptions: TestOptions.Regular.WithNullCheckingFeature(includeNonNullChecking: false));
+            comp.VerifyDiagnostics(
+                // (14,15): warning CS8604: Possible null reference argument for parameter 's' in 'void C.G(string s)'.
+                //             G(y); // warning
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "y").WithArguments("s", "void C.G(string s)").WithLocation(14, 15));
+        }
```

Comment:
Could you add a test with "default" literal if you have the feature in this branch and there isn't such a test already? 
If I remember correctly, it is represented the same as the default operator, in the bound tree, so I expect it will work fine (you already have tests for default(T)).